### PR TITLE
write directly to sys.stdout.buffer to avoid windows io encoding

### LINKED ISSUE #1381: flake8 crash on windows-latest with UnicodeEncodeError
*Please describe how you installed Flake8*

I using github actions run this check
```
name: Run Tests

on: push

jobs:
  test:
    runs-on: windows-latest

    steps:
    - uses: actions/checkout@v2
    - name: Set up Python 3.9
      uses: actions/setup-python@v2
      with:
        python-version: 3.9
    - name: Install dependencies
      run: |
        python -m pip install --upgrade pip
        pip install flake8
    - name: Additional info
      run: flake8 --bug-report
    - name: Lint with flake8
      run: |
        flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics
        flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics
```

(also checked on other version of python, same error)

And I get this tracelog

```
> Run `flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics`
0
Traceback (most recent call last):
  File "c:\hostedtoolcache\windows\python\3.9.6\x64\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "c:\hostedtoolcache\windows\python\3.9.6\x64\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\hostedtoolcache\windows\Python\3.9.6\x64\Scripts\flake8.exe\__main__.py", line 7, in <module>
  File "c:\hostedtoolcache\windows\python\3.9.6\x64\lib\site-packages\flake8\main\cli.py", line 22, in main
    app.run(argv)
  File "c:\hostedtoolcache\windows\python\3.9.6\x64\lib\site-packages\flake8\main\application.py", line 363, in run
    self._run(argv)
  File "c:\hostedtoolcache\windows\python\3.9.6\x64\lib\site-packages\flake8\main\application.py", line 352, in _run
    self.report()
  File "c:\hostedtoolcache\windows\python\3.9.6\x64\lib\site-packages\flake8\main\application.py", line 343, in report
    self.report_errors()
  File "c:\hostedtoolcache\windows\python\3.9.6\x64\lib\site-packages\flake8\main\application.py", line 298, in report_errors
    results = self.file_checker_manager.report()
  File "c:\hostedtoolcache\windows\python\3.9.6\x64\lib\site-packages\flake8\checker.py", line 261, in report
    results_reported += self._handle_results(filename, results)
  File "c:\hostedtoolcache\windows\python\3.9.6\x64\lib\site-packages\flake8\checker.py", line 159, in _handle_results
    reported_results_count += style_guide.handle_error(
  File "c:\hostedtoolcache\windows\python\3.9.6\x64\lib\site-packages\flake8\style_guide.py", line 424, in handle_error
    return guide.handle_error(
  File "c:\hostedtoolcache\windows\python\3.9.6\x64\lib\site-packages\flake8\style_guide.py", line 579, in handle_error
    self.formatter.handle(error)
  File "c:\hostedtoolcache\windows\python\3.9.6\x64\lib\site-packages\flake8\formatting\base.py", line 95, in handle
    self.write(line, source)
  File "c:\hostedtoolcache\windows\python\3.9.6\x64\lib\site-packages\flake8\formatting\base.py", line 206, in write
    self._write(line)
  File "c:\hostedtoolcache\windows\python\3.9.6\x64\lib\site-packages\flake8\formatting\base.py", line 189, in _write
    print(output, end=self.newline)
  File "c:\hostedtoolcache\windows\python\3.9.6\x64\lib\encodings\cp1252.py", line 19, in encode
    return codecs.charmap_encode(input,self.errors,encoding_table)[0]
UnicodeEncodeError: 'charmap' codec can't encode characters in position 36-40: character maps to <undefined>
````

*Please provide the exact, unmodified output of `flake8 --bug-report`*

```
{
  "dependencies": [],
  "platform": {
    "python_implementation": "CPython",
    "python_version": "3.9.6",
    "system": "Windows"
  },
  "plugins": [
    {
      "is_local": false,
      "plugin": "mccabe",
      "version": "0.6.1"
    },
    {
      "is_local": false,
      "plugin": "pycodestyle",
      "version": "2.7.0"
    },
    {
      "is_local": false,
      "plugin": "pyflakes",
      "version": "2.3.1"
    }
  ],
  "version": "3.9.2"
}
```

P.S. Using forked repository from [this commit](https://github.com/PerchunPak/PingerBot/tree/db64725bc964213fe2eb0783f23f69ef3236c5b4)

### PR DESCRIPTION
**Pull request by asottile**
resolves #1381

### CODE REVIEW

File: tests/unit/test_base_formatter.py [hunk 1]
```diff
@@ -130,21 +125,15 @@ def test_write_uses_an_output_file(tee):
     ]
 
 
-@mock.patch("flake8.formatting.base.print")
-def test_write_uses_print(print_function):
+def test_write_uses_print(capsys):
```
Comment (sigmavirus24) [inline tests/unit/test_base_formatter.py:128]: Should rename this too

Comment (asottile) [reply tests/unit/test_base_formatter.py:128]: oh yeah good point!

**Comment by HebaruSan** [comment]
Flake8 throws AttributeError on this line now. Any suggestions?

https://github.com/KSP-CKAN/NetKAN-Infra/pull/228/checks?check_run_id=3886107937

```
=================================== FAILURES ===================================
_________________________________ FLAKE8-check _________________________________
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/pluggy/_hooks.py:265: in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/pluggy/_manager.py:80: in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/_pytest/runner.py:170: in pytest_runtest_call
    raise e
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/_pytest/runner.py:162: in pytest_runtest_call
    item.runtest()
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/pytest_flake8.py:127: in runtest
    self.statistics)
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/py/_io/capture.py:150: in call
    res = func(*args, **kwargs)
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/pytest_flake8.py:222: in check_file
    app.report_errors()
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/flake8/main/application.py:309: in report_errors
    results = self.file_checker_manager.report()
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/flake8/checker.py:249: in report
    results_reported += self._handle_results(filename, results)
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/flake8/checker.py:161: in _handle_results
    physical_line=physical_line,
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/flake8/style_guide.py:430: in handle_error
    code, filename, line_number, column_number, text, physical_line
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/flake8/style_guide.py:581: in handle_error
    self.formatter.handle(error)
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/flake8/formatting/base.py:100: in handle
    self.write(line, source)
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/flake8/formatting/base.py:203: in write
    self._write(line)
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/flake8/formatting/base.py:187: in _write
    sys.stdout.buffer.write(output.encode() + self.newline.encode())
E   AttributeError: '_io.StringIO' object has no attribute 'buffer'
```

**Comment by HebaruSan** [comment]
Why the downthumbs on my previous comment, @asottile and @sigmavirus24? Line 187 of `base.py` is throwing an `AttributeError` when trying to access `sys.stdout.buffer`, and this pull request added that. If you already know why that is, would you mind sharing?

**Comment by asottile** [comment]
because you failed to look at pinned issues or do any due diligence and you're using an unsupported private api

**Comment by HebaruSan** [comment]
I'm not using _any_ API, I'm just running pytest with this in `pytest.ini`, to consolidate all of our Python tooling into a single output format:

```ini
[pytest]
addopts = -p no:cacheprovider --mypy --pylint --flake8
```

I've looked at the pinned issue. It says something about monkeypatching `sys.stdout`, and something else about `io.TextIOWrapper`, and then it was closed. I'm not monkeypatching anything, so that's not much help.

I guess the key takeaway here is that people are having flake8 break on them without themselves commiting whatever heinous technical sin so offends you here, because apparently pytest did it. I would suggest reconsidering how you are approaching that segment of your user base.

**Comment by asottile** [comment]
it's not pytest either, place your blame on pytest-flake8

also please adjust your attitude

**Pull request merged**